source/server/listener_manager_impl.cc

```diff
@@ -313,20 +313,50 @@ void ListenerImpl::setSocket(const Network::SocketSharedPtr& socket) {
 ListenerManagerImpl::ListenerManagerImpl(Instance& server,
                                          ListenerComponentFactory& listener_factory,
                                          WorkerFactory& worker_factory)
-    : server_(server), factory_(listener_factory), stats_(generateStats(server.stats())) {
+    : server_(server), factory_(listener_factory), stats_(generateStats(server.stats())),
+      config_tracker_entry_(server.admin().getConfigTracker().add(
+          "listeners", [this] { return dumpListenerConfigs(); })) {
   for (uint32_t i = 0; i < std::max(1U, server.options().concurrency()); i++) {
     workers_.emplace_back(worker_factory.createWorker());
   }
 }
 
+ProtobufTypes::MessagePtr ListenerManagerImpl::dumpListenerConfigs() {
+  auto config_dump = std::make_unique<envoy::admin::v2alpha::ListenersConfigDump>();
+  config_dump->set_version_info(lds_api_ != nullptr ? lds_api_->versionInfo() : "");
+  for (const auto& listener : active_listeners_) {
+    if (listener->blockRemove()) {
+      config_dump->mutable_static_listeners()->Add()->MergeFrom(listener->config());
+    } else {
+      auto& dynamic_listener = *config_dump->mutable_dynamic_active_listeners()->Add();
+      dynamic_listener.set_version_info(listener->versionInfo());
+      dynamic_listener.mutable_listener()->MergeFrom(listener->config());
+    }
+  }
+
+  for (const auto& listener : warming_listeners_) {
+    auto& dynamic_listener = *config_dump->mutable_dynamic_warming_listeners()->Add();
+    dynamic_listener.set_version_info(listener->versionInfo());
+    dynamic_listener.mutable_listener()->MergeFrom(listener->config());
+  }
+
+  for (const auto& listener : draining_listeners_) {
+    auto& dynamic_listener = *config_dump->mutable_dynamic_draining_listeners()->Add();
+    dynamic_listener.set_version_info(listener.listener_->versionInfo());
+    dynamic_listener.mutable_listener()->MergeFrom(listener.listener_->config());
+  }
```

Comment:
Can we factor out some of these copy loops in LDS/CDS?

Reply:
I thought about that, but the loops are subtly different, and I'm really not sure it's worth it to have a bunch of small functions here.